src/Task.cc

```diff
@@ -1534,18 +1565,100 @@ bool Task::set_debug_regs(const DebugRegs& regs) {
   for (size_t i = 0; i < regs.size(); ++i) {
     dr7.ctl.enable(i, BYTES_1, WATCH_EXEC);
   }
-  set_debug_reg(7, dr7.packed);
+  t->set_x86_debug_reg(7, dr7.packed);
 
   size_t index = 0;
   for (auto reg : regs) {
-    if (!set_debug_reg(index, reg.addr.as_int())) {
-      set_debug_reg(7, 0);
+    if (!t->set_x86_debug_reg(index, reg.addr.as_int())) {
+      t->set_x86_debug_reg(7, 0);
       return false;
     }
     dr7.ctl.enable(index, num_bytes_to_dr_len(reg.num_bytes), reg.type);
     ++index;
   }
-  return set_debug_reg(7, dr7.packed);
+  return t->set_x86_debug_reg(7, dr7.packed);
+}
+
+template <typename Arch>
+static bool set_debug_regs_arch(Task* t, const Task::DebugRegs& regs);
+template <> bool set_debug_regs_arch<X86Arch>(Task* t, const Task::DebugRegs& regs) {
+  return set_x86_debug_regs(t, regs);
+}
+template <> bool set_debug_regs_arch<X64Arch>(Task* t, const Task::DebugRegs& regs) {
+  return set_x86_debug_regs(t, regs);
+}
+
+static void query_max_bp_wp(Task* t, ssize_t* max_bp, ssize_t* max_wp) {
+  ARM64Arch::user_hwdebug_state bps;
+  ARM64Arch::user_hwdebug_state wps;
+  bool ok = t->get_aarch64_debug_regs(NT_ARM_HW_BREAK, &bps) &&
+            t->get_aarch64_debug_regs(NT_ARM_HW_WATCH, &wps);
+  ASSERT(t, ok);
+  *max_bp = bps.dbg_info & 0xff;
+  *max_wp = wps.dbg_info & 0xff;
+}
+
+template <> bool set_debug_regs_arch<ARM64Arch>(Task* t, const Task::DebugRegs& regs) {
+  ARM64Arch::user_hwdebug_state bps;
+  ARM64Arch::user_hwdebug_state wps;
+  memset(&bps, 0, sizeof(bps));
+  memset(&wps, 0, sizeof(wps));
+
+  static ssize_t max_bp = -1;
+  static ssize_t max_wp = -1;
+  if (max_bp == -1) {
+    query_max_bp_wp(t, &max_bp, &max_wp);
+  }
+
+  // Having at least one of each is architecturally guaranteed
+  ASSERT(t, max_bp >= 1 && max_wp >= 1);
+
+  ssize_t cur_bp = 0;
+  ssize_t cur_wp = 0;
+  for (auto reg : regs) {
+    ARM64Arch::hw_bp* bp = nullptr;
+    if (reg.type == WATCH_EXEC) {
+      if (cur_bp == max_bp) {
+        return false;
+      }
+      bp = &bps.dbg_regs[cur_bp++];
+    } else {
+      if (cur_wp == max_wp) {
+        return false;
+      }
+      bp = &wps.dbg_regs[cur_wp++];
+    }
+    ARM64Arch::hw_breakpoint_ctrl ctrl;
+    memset(&ctrl, 0, sizeof(ctrl));
+    switch (reg.type) {
+      case WATCH_EXEC:
+        ctrl.type = ARM_WATCH_EXEC;
+        break;
+      case WATCH_WRITE:
+        ctrl.type = ARM_WATCH_WRITE;
+        break;
+      case WATCH_READWRITE:
+        ctrl.type = ARM_WATCH_READWRITE;
+        break;
+    }
+    ctrl.enabled = 1;
+    ctrl.priv = ARM_PRIV_EL0;
+	  uintptr_t off = (uintptr_t)reg.addr.as_int() % 8;
```

Comment:
1. What happened to the indentation here?
2. Does gdb automatically break things that cross the alignment boundary here for us? i.e. it's not going to give us 0xABCD4 with a watchpoint length of 8?

Reply:
Re indentation, not sure. Hmm, not sure. I may have accidentally introduced some tabs or something by editing with vim.

As for the alignment thing, good question. I meant to check that, but forgot.

Reply:
I'm holding off on merging this solely because of this. No rush, but if you're waiting on me I wanted to let you know why this is sitting here.

Reply:
Nope fair, just haven't gotten around to it yet.

Reply:
Looks like the answer is no, it does not split. I'll adjust the code to do that here.

Reply:
Actually looks like I was wrong. It does split, but for some reason the watchpoint doesn't get hit. Let me investigate.

Reply:
Turns out I was just testing wrong. This is fine as is, but I'll push the commit to split the watchpoints anyway, since it's small and should provide some extra assurance for any future changes in this GDB behavior.